### ISSUE TITLE
Add an overridable logging function

### ISSUE THREAD

**Issue by zainhoda** [vanna-ai/vanna#176]
Instead of using hardcoded print statements, there should be a logging function in the base class. That logging function in particular should be called before submitting the prompt so that you can log what the fully constructed prompt looks like.

By default, the log function in the base class should just be a print statement. If a user wants to send logs to a specific destination, they can just override the log function.

**Comment by justbee007** [comment]
Hi I am new to contributing for open source and would like to contribute to this project.

**Comment by pedro-geraghty-sw** [comment]
It would be nice to be able to log metrics on times response and token consumption and other related stuff

**Comment by justbee007** [comment]
@pedro-geraghty-sw are you gonna work on it?. I thought of fixing this issue.

**Comment by pedro-geraghty-sw** [comment]
Go ahead

**Comment by NickCrews** [comment]
Thoughts:
- The log function should take the same severity levels as the building logging module as an optional kwarg. Default is INFO?
- If we're smart, it could also support structured logging eg structlog.

**Comment by zainhoda** [comment]
> Thoughts:
> 
> * The log function should take the same severity levels as the building logging module as an optional kwarg. Default is INFO?
> * If we're smart, it could also support structured logging eg structlog.

Agreed. Also, I think it would be good for `generate_sql` to start with a uuid as a trace_id that gets passed into all the constituent functions so that logs for the same `generate_sql` request can easily be connected

**Comment by justbee007** [comment]
You can assign this ticket to me. 
Thank You

**Comment by andreped** [comment]
What is the status on this?

I can make a PR for this today, as I need it for one of my projects very urgently.

I have made similar logging solutions for framework before, so should not be that difficult. I can tag you in, @justbee007, if you wish to contribute to the PR.

**Comment by justbee007** [comment]
@andreped Yes Please you can make a PR and tag me in. We can work on this together.

**Comment by andreped** [comment]
> @andreped Yes Please you can make a PR and tag me in. We can work on this together.

Drafted a PR: https://github.com/vanna-ai/vanna/pull/234

Feel free to suggest improvements. I already hinted at some potential suggestions in the PR comments.

**Comment by talrejanikhil** [comment]
Was this PR merged and released?

**Comment by andreped** [comment]
> Was this PR merged and released?

@talrejanikhil No, but I believe @zainhoda implemented a different solution. Not sure though.
